File: LeetCode/0270.Closest-Binary-Search-Tree-Value/Closest-Binary-Search-Tree-Value.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */

public class Solution {
    /**
     * @param root: the given BST
     * @param target: the given target
     * @return: the value in the BST that is closest to the target
     */
    public int closestValue(TreeNode root, double target) {
        // write your code here
        if(root == null) return -1;
        TreeNode lower = root;
        TreeNode upper = root;
        while(root != null){
            if(target == root.val){
                return root.val;
            }
            else if(target > root.val){
                lower = root;
                root = root.right;
            }
            else{
                upper = root;
                root = root.left;
            }
        }
        if(Math.abs(upper.val - target) > Math.abs(target - lower.val)){
            return lower.val;
        }else{
            return upper.val;
        }
    }
}

File: LeetCode/0270.Closest-Binary-Search-Tree-Value/Closest-Binary-Search-Tree-Value.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def closestValue(self, root, target):
        """
        :type root: TreeNode
        :type target: float
        :rtype: int
        """
        return self.closestValue2(root, target)
    
    def closestValue2(self, root, target):
        res = root.val
        node = root
        while node:
            if abs(node.val - target) <= abs(res - target):
                res = node.val
            node = node.left if target < node.val else node.right
        return res
            
        
    def closestValue1(self, root, target):
        res, diff = None, float('inf')
        stack = []
        node = root
        while stack or node:
            while node:
                stack.append(node)
                node = node.left
            cur = stack.pop()
            if abs(cur.val - target) < diff:
                diff = abs(cur.val - target)
                res = cur.val
            else:
                break
            node = cur.right
        return res
